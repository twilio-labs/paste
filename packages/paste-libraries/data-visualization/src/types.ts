import * as Highcharts from "highcharts";

type SupportedCharts = "line" | "column" | "area";

/**
 * The object structure required for use as a `series` prop in a chart with the type `line`.
 */
export type LineSeries = Pick<
  Highcharts.SeriesLineOptions,
  "data" | "color" | "accessibility" | "dashStyle" | "lineWidth" | "marker" | "name"
>;

/**
 * The object structure required for use as a `series` prop in a chart with the type `area`.
 */
export type AreaSeries = Pick<Highcharts.SeriesAreaOptions, "data" | "color" | "accessibility" | "name">;

/**
 * The object structure required for use as a `series` prop in a chart with the type `column`.
 */
export type ColumnSeries = Pick<Highcharts.SeriesColumnOptions, "data" | "color" | "accessibility" | "events" | "name">;

export type ChartConfig = Pick<Highcharts.ChartOptions, "animation" | "height" | "width">;

type ChartAccessibilityConfig = Pick<
  Highcharts.AccessibilityOptions,
  "description" | "point" | "linkedDescription" | "series"
>;

interface TitleConfig extends Pick<Highcharts.TitleOptions, "text"> {
  hide?: boolean;
}

export interface BaseChartOptions {
  /**
   * Sets metadata for the chart such as whether it is animated, the height and width of the chart, and other chart options.
   */
  chart?: ChartConfig;
  /**
   * Sets global accessiblity options such as formatting the point and series text that is read to users when they focus on elements.
   */
  accessibility?: ChartAccessibilityConfig;
  /**
   * Configure the title of the chart. Optionall hide the title too. The purpose of setting text but hiding the title allows the
   * title to still be included in any chart descritions read by screen readers.
   */
  title?: TitleConfig;
  /**
   * Configure the sub title of the chart. Optionall hide the sub title too. The purpose of setting text but hiding the sub title allows the
   * sub title to still be included in any chart descritions read by screen readers.
   */
  subtitle?: TitleConfig;
  /**
   * The type of chart to render. This will determine what other chart specfic options are available.
   */
  type: SupportedCharts;
  /**
   * Render the X axis as a datetime.
   */
  isXTimeAxis?: boolean;
  /**
   * Whether to show the native legend provided by Highcharts.
   */
  showLegend?: boolean;
  /**
   * The type of zooming to enable on the chart. This will determine whether users can zoom in the x, y or both axis.
   */
  zoomingType?: Highcharts.ChartOptions["zoomType"];
  /**
   * The type of panning to enable on the chart allowing users to move the region currently zoomed in on.
   * This will determine whether users can pan in the x, y or both axis.
   */
  panningType?: Highcharts.ChartOptions["zoomType"];
  /**
   * A function that formats the point text for Highcharts native tooltips.
   */
  pointFormatter?: (point: Highcharts.Point) => string;
  /**
   * Configure the y axis of the chart. Optionall hide the y axis too. The purpose of setting text but hiding the y axis allows the
   * y axis to still be included in any chart descritions read by screen readers.
   */
  yAxisTitle?: TitleConfig;
  /**
   * Configure the x axis of the chart. Optionall hide the x axis too. The purpose of setting text but hiding the x axis allows the
   * x axis to still be included in any chart descritions read by screen readers.
   */
  xAxisTitle?: TitleConfig;
  /**
   * Controls the display of teh watermark in the bottom right of the chart accrediting Highcharts. Not needed for projects with a license and disbaled by default.
   *
   * @default false
   */
  enableCredits?: boolean;
  /**
   * Controls whether the chart is animated.
   *
   * @default true
   */
  isAnimated?: boolean;
  /**
   * The data to be displayed on the x axis labels. This is useful when you do not have the x plot data in your series but need to set the tick labels on teh x axis.
   */
  xAxisCategories?: Highcharts.XAxisOptions["categories"];
}

export interface LineChartConfig extends BaseChartOptions {
  /**
   * The series data for a line chart.
   */
  series: LineSeries[];
  /**
   * Specifies that this is a line chart.
   */
  type: "line";
}

export interface ColumnChartConfig extends BaseChartOptions {
  /**
   * The series data for a column chart.
   */
  series: ColumnSeries[];
  /**
   * Specifies that this is a column chart.
   */
  type: "column";
  /**
   * The type of stacking to use for the column chart, whether to show percentage stacked or normal.
   */
  stackingType?: "normal" | "percent";
  /**
   * Show the number value for the column on top of the section. Works for regular and stacked variants.
   */
  showDataLabels?: boolean;
}

export interface AreaChartConfig extends BaseChartOptions {
  /**
   * The series data for an area chart.
   */
  series: AreaSeries[];
  /**
   * Specifies that this is an area chart.
   */
  type: "area";
}

export type ChartTypeOptions = LineChartConfig | ColumnChartConfig | AreaChartConfig;
