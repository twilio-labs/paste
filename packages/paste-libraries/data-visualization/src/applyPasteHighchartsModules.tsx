import Highcharts from "highcharts";

/**
 * A hook to apply any Highcharts modules iwth accessibility being mandatory.
 */
export const applyPasteHighchartsModules = (
  highcharts: typeof Highcharts,
  accessibilityModule: (highcharts: typeof Highcharts) => void,
  ...rest: Array<(highcharts: typeof Highcharts) => void>
): void => {
  if (highcharts === undefined) {
    throw new Error("[applyPasteHighchartsModules]: Must provide highcharts into this function.");
  }

  if (accessibilityModule === undefined) {
    throw new Error(
      '[applyPasteHighchartsModules]: Must provide accessibility module into this function. You can do this using `import HighchartsAccessibility Module from "highcharts/modules/accessibility"`',
    );
  }

  accessibilityModule(highcharts);
  rest.forEach((module) => module(highcharts));
};
