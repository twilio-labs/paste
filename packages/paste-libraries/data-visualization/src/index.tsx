export { usePasteHighchartsTheme } from "./usePasteHighchartsTheme";
export { applyPasteHighchartsModules } from "./applyPasteHighchartsModules";
export { transformToHighchartsOptions } from "./transformers";
export type {
  BaseChartOptions,
  ChartTypeOptions,
  ChartConfig,
  AreaChartConfig,
  AreaSeries,
  ColumnChartConfig,
  ColumnSeries,
  LineChartConfig,
  LineSeries,
} from "./types";
