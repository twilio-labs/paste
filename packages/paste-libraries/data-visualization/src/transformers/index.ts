import { ChartTypeOptions, ColumnChartConfig, LineChartConfig } from "../types";
import { handleBaseChartOptionsFormatting } from "./baseFormatter";
import { handleColumnOptions } from "./columnChart";
import { handleLineOptions } from "./lineChart";

const handleDefaultOptions = ({ chart, series, type, ...baseChartProps }: ChartTypeOptions): Highcharts.Options => {
  return {
    // @ts-ignore mentions incompatible types but is correct
    series: series.map((s) => ({ ...s, type })),
    ...handleBaseChartOptionsFormatting({ ...baseChartProps, type }),
  };
};

export const transformToHighchartsOptions = (options: ChartTypeOptions): Highcharts.Options => {
  switch (options.type) {
    case "line":
      return handleLineOptions(options as LineChartConfig);
    //   case "area":
    //     return handleAreaOptions(options as AreaChartConfig);
    case "column":
      return handleColumnOptions(options as ColumnChartConfig);
    default: {
      return handleDefaultOptions(options);
    }
  }
};
