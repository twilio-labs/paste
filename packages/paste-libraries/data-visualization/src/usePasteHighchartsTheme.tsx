import * as React from 'react';
import {ThemeContext} from '@twilio-paste/styling-library';
import type {ThemeShape} from '@twilio-paste/theme';
import type Highcharts from 'highcharts';

const merge = require('deepmerge');

/**
 * A hook to apply the Paste theme to Highcharts options.
 *
 * @param {Highcharts.Options} options - options to be provided into Highcharts for rendering.
 * @returns {Highcharts.Options} optionsWithPasteTheme - Passed options with deepmerged Paste Theme
 */
export const usePasteHighchartsTheme = (options: Highcharts.Options): Highcharts.Options => {
  if (options == null) {
    throw new Error(
      `[usePasteHighchartsTheme]: Must provide highcharts options into this function. Options are deepmerged against the theme and returned for usage.`
    );
  }

  const context = React.useContext(ThemeContext) as ThemeShape;
  if (!context) {
    throw new Error(
      '[usePasteHighchartsTheme]: must be used within the @twilio-paste/theme provider. https://paste.twilio.design/introduction/for-engineers/manual-installation#setting-up-the-theme-provider'
    );
  }

  /*
   * context.dataVisualization returns the keys sorted alphabetically, which is a problem because:
   * -1 comes before -10 comes before -2.  We want -1, -2, ... -10 (numerical sorting).
   * For this reason we need to sort the keys numerically before retrieving their values.
   */
  const colors: string[] = React.useMemo(() => {
    const collator = new Intl.Collator(undefined, {numeric: true, sensitivity: 'base'});
    const tokenNames = Object.keys(context.dataVisualization) as Array<keyof ThemeShape['dataVisualization']>;
    const sortedTokenNames = tokenNames.sort((a, b) => collator.compare(a, b));

    // Now we can get the color array in the correct order
    return sortedTokenNames.map((name) => context.dataVisualization[name]);
  }, [context]);

  return merge(options, {
    colors,
    chart: {
      backgroundColor: context.backgroundColors.colorBackgroundBody,
    },
    title: {
      style: {
        color: context.textColors.colorText,
        font: `${context.fontWeights.fontWeightBold} ${context.fontSizes.fontSize40} ${context.fonts.fontFamilyText}`,
      },
    },
    subtitle: {
      style: {
        color: context.textColors.colorTextWeak,
        font: `${context.fontSizes.fontSize20} ${context.fonts.fontFamilyText}`,
      },
    },
    legend: {
      itemStyle: {
        font: `${context.fontSizes.fontSize10} ${context.fonts.fontFamilyText}`,
        color: context.textColors.colorText,
      },
      itemHoverStyle: {
        color: context.textColors.colorTextWeak,
      },
      title: {
        style: {
          color: context.textColors.colorText,
        },
      },
    },
    xAxis: {
      lineColor: context.textColors.colorText,
      tickColor: context.textColors.colorText,
      labels: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightMedium} ${context.fontSizes.fontSize10} ${context.fonts.fontFamilyText}`,
        },
      },
      title: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightSemibold} ${context.fontSizes.fontSize30} ${context.fonts.fontFamilyText}`,
        },
      },
    },
    yAxis: {
      labels: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightMedium} ${context.fontSizes.fontSize10} ${context.fonts.fontFamilyText}`,
        },
      },
      title: {
        style: {
          color: context.textColors.colorText,
          font: `${context.fontWeights.fontWeightSemibold} ${context.fontSizes.fontSize30} ${context.fonts.fontFamilyText}`,
        },
      },
    },
    labels: {
      style: {
        color: context.textColors.colorText,
      },
    },
    tooltip: {
      backgroundColor: context.backgroundColors.colorBackground,
      style: {
        color: context.textColors.colorText,
      },
    },
    credits: {
      style: {
        color: context.textColors.colorTextWeaker,
      },
    },
    plotOptions: {
      series: {
        borderColor: 'transparent',
      },
    },
  });
};
