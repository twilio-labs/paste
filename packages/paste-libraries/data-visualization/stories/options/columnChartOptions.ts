// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck copied from official highcharts docs
import type Highcharts from 'highcharts';

export const columnChartOptions: Highcharts.Options = {
  chart: {
    type: 'column',
  },
  title: {
    text: 'Monthly Average Rainfall',
  },
  subtitle: {
    text: 'Source: WorldClimate.com',
  },
  xAxis: {
    categories: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
    crosshair: true,
  },
  yAxis: {
    min: 0,
    title: {
      text: 'Rainfall (mm)',
    },
  },
  series: [
    {
      name: 'Tokyo',
      data: [49.9, 71.5, 106.4, 129.2, 144, 176, 135.6, 148.5, 216.4, 194.1, 95.6, 54.4],
    },
    {
      name: 'New York',
      data: [83.6, 78.8, 98.5, 93.4, 106, 84.5, 105, 104.3, 91.2, 83.5, 106.6, 92.3],
    },
    {
      name: 'London',
      data: [48.9, 38.8, 39.3, 41.4, 47.1, 48.3, 59.2, 59.6, 52.4, 65.2, 59.3, 51.2],
    },
    {
      name: 'Berlin',
      data: [42.4, 33.2, 34.5, 39.7, 52.6, 75.5, 57.4, 60.4, 47.6, 39.1, 46.8, 51.1],
    },
  ],
};
