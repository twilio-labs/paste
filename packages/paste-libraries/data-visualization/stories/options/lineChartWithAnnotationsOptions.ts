// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck copied from official highcharts docs
import Highcharts from 'highcharts';
import {elevationData} from '../data/elevationData';

export const lineChartWithAnnotationsOptions: Highcharts.Options = {
  chart: {
    type: 'area',
    zoomType: 'x',
    panKey: 'shift',
    scrollablePlotArea: {
      minWidth: 600,
    },
  },

  caption: {
    text: 'This chart uses the Highcharts Annotations feature to place labels at various points of interest. The labels are responsive and will be hidden to avoid overlap on small screens.',
  },

  title: {
    text: '2017 Tour de France Stage 8: Dole - Station des Rousses',
  },

  accessibility: {
    description:
      'This line chart uses the Highcharts Annotations feature to place labels at various points of interest. The labels are responsive and will be hidden to avoid overlap on small screens. Image description: An annotated line chart illustrates the 8th stage of the 2017 Tour de France cycling race from the start point in Dole to the finish line at Station des Rousses. Altitude is plotted on the Y-axis, and distance is plotted on the X-axis. The line graph is interactive, and the user can trace the altitude level along the stage. The graph is shaded below the data line to visualize the mountainous altitudes encountered on the 187.5-kilometre stage. The three largest climbs are highlighted at Col de la Joux, Côte de Viry and the final 11.7-kilometer, 6.4% gradient climb to Montée de la Combe de Laisia Les Molunes which peaks at 1200 meters above sea level. The stage passes through the villages of Arbois, Montrond, Bonlieu, Chassal and Saint-Claude along the route.',
    landmarkVerbosity: 'one',
  },

  lang: {
    accessibility: {
      screenReaderSection: {
        annotations: {
          descriptionNoPoints:
            '{annotationText}, at distance {annotation.options.point.x}km, elevation {annotation.options.point.y} meters.',
        },
      },
    },
  },

  credits: {
    enabled: false,
  },

  annotations: [
    {
      draggable: '',
      labelOptions: {
        backgroundColor: 'rgba(255,255,255,0.5)',
        verticalAlign: 'top',
        y: 15,
      },
      labels: [
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 27.98,
            y: 255,
          },
          text: 'Arbois',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 45.5,
            y: 611,
          },
          text: 'Montrond',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 63,
            y: 651,
          },
          text: 'Mont-sur-Monnet',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 84,
            y: 789,
          },
          x: -10,
          text: 'Bonlieu',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 129.5,
            y: 382,
          },
          text: 'Chassal',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 159,
            y: 443,
          },
          text: 'Saint-Claude',
        },
      ],
    },
    {
      draggable: '',
      labels: [
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 101.44,
            y: 1026,
          },
          x: -30,
          text: 'Col de la Joux',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 138.5,
            y: 748,
          },
          text: 'Côte de Viry',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 176.4,
            y: 1202,
          },
          text: 'Montée de la Combe <br>de Laisia Les Molunes',
        },
      ],
    },
    {
      draggable: '',
      labelOptions: {
        shape: 'connector',
        align: 'right',
        justify: false,
        crop: true,
        style: {
          fontSize: '0.8em',
          textOutline: '1px white',
        },
      },
      labels: [
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 96.2,
            y: 783,
          },
          text: '6.1 km climb <br>4.6% on avg.',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 134.5,
            y: 540,
          },
          text: '7.6 km climb <br>5.2% on avg.',
        },
        {
          point: {
            xAxis: 0,
            yAxis: 0,
            x: 172.2,
            y: 925,
          },
          text: '11.7 km climb <br>6.4% on avg.',
        },
      ],
    },
  ],

  xAxis: {
    labels: {
      format: '{value} km',
    },
    minRange: 5,
    title: {
      text: 'Distance',
    },
    accessibility: {
      rangeDescription: 'Range: 0 to 187.8km.',
    },
  },

  yAxis: {
    startOnTick: true,
    endOnTick: false,
    maxPadding: 0.35,
    title: {
      text: null,
    },
    labels: {
      format: '{value} m',
    },
    accessibility: {
      description: 'Elevation',
      rangeDescription: 'Range: 0 to 1,553 meters',
    },
  },

  tooltip: {
    headerFormat: 'Distance: {point.x:.1f} km<br>',
    pointFormat: '{point.y} m a. s. l.',
    shared: true,
  },

  legend: {
    enabled: false,
  },

  series: [
    {
      data: elevationData,
      lineColor: Highcharts.getOptions().colors[1],
      color: Highcharts.getOptions().colors[2],
      fillOpacity: 0.5,
      name: 'Elevation',
      marker: {
        enabled: false,
      },
      threshold: null,
    },
  ],
};
