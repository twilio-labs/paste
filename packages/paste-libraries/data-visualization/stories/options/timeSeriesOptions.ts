// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck copied from official highcharts docs
import Highcharts from 'highcharts';
import {usdEurData} from '../data/usdEurData';

export const timeSeriesOptions: Highcharts.Options = {
  chart: {
    zoomType: 'x',
  },
  title: {
    text: 'USD to EUR exchange rate over time',
  },
  subtitle: {
    text:
      document.ontouchstart === undefined ? 'Click and drag in the plot area to zoom in' : 'Pinch the chart to zoom in',
  },
  xAxis: {
    type: 'datetime',
  },
  yAxis: {
    title: {
      text: 'Exchange rate',
    },
  },
  legend: {
    enabled: false,
  },
  plotOptions: {
    area: {
      fillColor: {
        linearGradient: {
          x1: 0,
          y1: 0,
          x2: 0,
          y2: 1,
        },
        stops: [
          [0, Highcharts.getOptions().colors[0]],
          [1, Highcharts.color(Highcharts.getOptions().colors[0]).setOpacity(0).get('rgba')],
        ],
      },
      marker: {
        radius: 2,
      },
      lineWidth: 1,
      states: {
        hover: {
          lineWidth: 1,
        },
      },
      threshold: null,
    },
  },
  series: [
    {
      type: 'area',
      name: 'USD to EUR',
      data: usdEurData,
    },
  ],
};
