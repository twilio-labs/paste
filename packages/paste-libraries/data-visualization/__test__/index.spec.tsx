import * as React from 'react';
import {render} from '@testing-library/react';
import {Theme} from '@twilio-paste/theme';
import {usePasteHighchartsTheme} from '../src';
import {lineChartOptions} from '../stories/options/lineChartOptions';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const TestComponent: React.FC<{options?: {[key: string]: any}}> = ({options = {}}) => {
  const chartOptions = usePasteHighchartsTheme(options);
  return <span data-testid="test">{JSON.stringify(chartOptions)}</span>;
};

describe('usePasteHighChartsTheme hook', () => {
  it('throws an error when no arguments are passed', () => {
    expect(usePasteHighchartsTheme).toThrowError();
  });

  it('throws an error when not used within the Paste Theme.Provider', () => {
    expect(usePasteHighchartsTheme).toThrowError();
  });

  it('should return the theme options object when an empty object is passed', () => {
    const {getByTestId} = render(
      <Theme.Provider theme="default">
        <TestComponent />
      </Theme.Provider>
    );
    const content = JSON.parse(getByTestId('test').innerHTML);
    expect(content).toMatchSnapshot();
  });

  it('should return a deepmerged options object when options are passed', () => {
    const {getByTestId} = render(
      <Theme.Provider theme="default">
        <TestComponent options={lineChartOptions} />
      </Theme.Provider>
    );
    const content = JSON.parse(getByTestId('test').innerHTML);
    expect(content).toMatchSnapshot();
  });
});
