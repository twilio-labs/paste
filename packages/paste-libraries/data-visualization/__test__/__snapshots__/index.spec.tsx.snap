// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePasteHighChartsTheme hook should return a deepmerged options object when options are passed 1`] = `
Object {
  "chart": Object {
    "backgroundColor": "rgb(255, 255, 255)",
  },
  "colors": Array [
    "rgb(0, 20, 137)",
    "rgb(14, 124, 58)",
    "rgb(13, 58, 31)",
    "rgb(0, 140, 255)",
    "rgb(57, 71, 98)",
    "rgb(166, 127, 227)",
    "rgb(109, 46, 209)",
    "rgb(136, 145, 170)",
    "rgb(117, 12, 12)",
    "rgb(235, 86, 86)",
  ],
  "credits": Object {
    "style": Object {
      "color": "rgb(174, 178, 193)",
    },
  },
  "labels": Object {
    "style": Object {
      "color": "rgb(18, 28, 45)",
    },
  },
  "legend": Object {
    "itemHoverStyle": Object {
      "color": "rgb(96, 107, 133)",
    },
    "itemStyle": Object {
      "color": "rgb(18, 28, 45)",
      "font": "0.625rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
    },
    "title": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
      },
    },
  },
  "plotOptions": Object {
    "series": Object {
      "borderColor": "transparent",
    },
  },
  "series": Array [
    Object {
      "data": Array [
        43934,
        52503,
        57177,
        69658,
        97031,
        119931,
        137133,
        154175,
      ],
      "name": "Installation",
    },
    Object {
      "data": Array [
        24916,
        24064,
        29742,
        29851,
        32490,
        30282,
        38121,
        40434,
      ],
      "name": "Manufacturing",
    },
    Object {
      "data": Array [
        11744,
        17722,
        16005,
        19771,
        20185,
        24377,
        32147,
        39387,
      ],
      "name": "Sales &amp; Distribution",
    },
    Object {
      "data": Array [
        null,
        null,
        7988,
        12169,
        15112,
        22452,
        34400,
        34227,
      ],
      "name": "Project Development",
    },
    Object {
      "data": Array [
        12908,
        5948,
        8105,
        11248,
        8989,
        11816,
        18274,
        18111,
      ],
      "name": "Other",
    },
  ],
  "subtitle": Object {
    "style": Object {
      "color": "rgb(96, 107, 133)",
      "font": "0.75rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
    },
    "text": "Source: thesolarfoundation.com",
  },
  "title": Object {
    "style": Object {
      "color": "rgb(18, 28, 45)",
      "font": "700 1rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
    },
    "text": "Solar Employment Growth by Sector, 2010-2016",
  },
  "tooltip": Object {
    "backgroundColor": "rgb(244, 244, 246)",
    "style": Object {
      "color": "rgb(18, 28, 45)",
    },
  },
  "xAxis": Object {
    "labels": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "500 0.625rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
    "lineColor": "rgb(18, 28, 45)",
    "tickColor": "rgb(18, 28, 45)",
    "title": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "600 0.875rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
  },
  "yAxis": Object {
    "labels": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "500 0.625rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
    "title": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "600 0.875rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
  },
}
`;

exports[`usePasteHighChartsTheme hook should return the theme options object when an empty object is passed 1`] = `
Object {
  "chart": Object {
    "backgroundColor": "rgb(255, 255, 255)",
  },
  "colors": Array [
    "rgb(0, 20, 137)",
    "rgb(14, 124, 58)",
    "rgb(13, 58, 31)",
    "rgb(0, 140, 255)",
    "rgb(57, 71, 98)",
    "rgb(166, 127, 227)",
    "rgb(109, 46, 209)",
    "rgb(136, 145, 170)",
    "rgb(117, 12, 12)",
    "rgb(235, 86, 86)",
  ],
  "credits": Object {
    "style": Object {
      "color": "rgb(174, 178, 193)",
    },
  },
  "labels": Object {
    "style": Object {
      "color": "rgb(18, 28, 45)",
    },
  },
  "legend": Object {
    "itemHoverStyle": Object {
      "color": "rgb(96, 107, 133)",
    },
    "itemStyle": Object {
      "color": "rgb(18, 28, 45)",
      "font": "0.625rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
    },
    "title": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
      },
    },
  },
  "plotOptions": Object {
    "series": Object {
      "borderColor": "transparent",
    },
  },
  "subtitle": Object {
    "style": Object {
      "color": "rgb(96, 107, 133)",
      "font": "0.75rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
    },
  },
  "title": Object {
    "style": Object {
      "color": "rgb(18, 28, 45)",
      "font": "700 1rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
    },
  },
  "tooltip": Object {
    "backgroundColor": "rgb(244, 244, 246)",
    "style": Object {
      "color": "rgb(18, 28, 45)",
    },
  },
  "xAxis": Object {
    "labels": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "500 0.625rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
    "lineColor": "rgb(18, 28, 45)",
    "tickColor": "rgb(18, 28, 45)",
    "title": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "600 0.875rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
  },
  "yAxis": Object {
    "labels": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "500 0.625rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
    "title": Object {
      "style": Object {
        "color": "rgb(18, 28, 45)",
        "font": "600 0.875rem 'Inter var experimental', 'Inter var', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif",
      },
    },
  },
}
`;
